src/gadgets/k_mix.rs

```diff
@@ -199,48 +345,192 @@ mod tests {
+        let mid_vars = mid_com.commit(&mut verifier);
         let output_vars = output_com.commit(&mut verifier);
 
-        let mut verifier_cs = verifier.finalize_inputs();
-
         // Verifier adds constraints to the constraint system
-        assert!(fill_cs(&mut verifier_cs, input_vars, inter_vars, output_vars).is_ok());
+        assert!(call_mix_gadget(&mut verifier, &input_vars, &mid_vars, &output_vars).is_ok());
+
+        Ok(verifier.verify(&proof)?)
+    }
+
+    // Note: the output vectors for order_by_flavor does not have to be in a particular order,
+    // they just has to be grouped by flavor. Thus, it is possible to make a valid change to
+    // order_by_flavor but break the tests.
+    #[test]
+    fn order_by_flavor_test() {
+        let pc_gens = PedersenGens::default();
+        let bp_gens = BulletproofGens::new(128, 1);
+        let mut transcript = Transcript::new(b"OrderByFlavorTest");
+        let mut prover_cs = Prover::new(&bp_gens, &pc_gens, &mut transcript);
+
+        // k = 1
+        assert_eq!(
+            order_by_flavor(&vec![yuan(1)], &mut prover_cs).unwrap().1,
+            vec![yuan(1)]
+        );
+        // k = 2
+        assert_eq!(
+            order_by_flavor(&vec![yuan(1), yuan(2)], &mut prover_cs)
+                .unwrap()
+                .1,
+            vec![yuan(1), yuan(2)]
+        );
+        assert_eq!(
+            order_by_flavor(&vec![yuan(1), peso(2)], &mut prover_cs)
+                .unwrap()
+                .1,
+            vec![yuan(1), peso(2)]
+        );
+        // k = 3
+        assert_eq!(
+            order_by_flavor(&vec![yuan(1), peso(3), yuan(2)], &mut prover_cs)
+                .unwrap()
+                .1,
+            vec![yuan(1), yuan(2), peso(3)]
+        );
+        // k = 4
+        assert_eq!(
+            order_by_flavor(&vec![yuan(1), peso(3), yuan(2), peso(4)], &mut prover_cs)
+                .unwrap()
+                .1,
+            vec![yuan(1), yuan(2), peso(3), peso(4)]
+        );
+        assert_eq!(
+            order_by_flavor(&vec![yuan(1), peso(3), peso(4), yuan(2)], &mut prover_cs)
+                .unwrap()
+                .1,
+            vec![yuan(1), yuan(2), peso(4), peso(3)]
+        );
+        assert_eq!(
+            order_by_flavor(&vec![yuan(1), peso(3), zero(), yuan(2)], &mut prover_cs)
+                .unwrap()
+                .1,
+            vec![yuan(1), yuan(2), zero(), peso(3)]
+        );
+        assert_eq!(
+            order_by_flavor(&vec![yuan(1), yuan(2), yuan(3), yuan(4)], &mut prover_cs)
+                .unwrap()
+                .1,
+            vec![yuan(1), yuan(4), yuan(3), yuan(2)]
+        );
+        // k = 5
+        assert_eq!(
+            order_by_flavor(
+                &vec![yuan(1), yuan(2), yuan(3), yuan(4), yuan(5)],
+                &mut prover_cs
+            )
+            .unwrap()
+            .1,
+            vec![yuan(1), yuan(5), yuan(4), yuan(3), yuan(2)]
```

Comment:
Seems like the sorter reverses the list of the same-flavor items except for the first item. Can this be explained? Would the code that maintains the order become more complicated? Just curious.

Reply:
Got it, makes sense!